src/ring_theory/valuation/basic.lean

```diff
@@ -0,0 +1,392 @@
+
+lemma ne_zero (h : v₁.is_equiv v₂) {r : R} :
+  v₁ r ≠ 0 ↔ v₂ r ≠ 0 :=
+begin
+  have : v₁ r ≠ v₁ 0 ↔ v₂ r ≠ v₂ 0 := not_iff_not_of_iff h.val_eq,
+  rwa [v₁.map_zero, v₂.map_zero] at this,
+end
+
+end is_equiv -- end of namespace
+
+lemma is_equiv_of_map_strict_mono [ring R] {v : valuation R Γ₀}
+  (f : Γ₀ →* Γ'₀) (h₀ : f 0 = 0) (H : strict_mono f) :
+  is_equiv (v.map f h₀ (H.monotone)) v :=
+λ x y, ⟨H.le_iff_le.mp, λ h, H.monotone h⟩
+
+lemma is_equiv_of_val_le_one {K : Type*} [division_ring K]
+  (v : valuation K Γ₀) (v' : valuation K Γ'₀) (h : ∀ {x:K}, v x ≤ 1 ↔ v' x ≤ 1) :
+  v.is_equiv v' :=
+begin
+  intros x y,
+  by_cases hy : y = 0, { simp [hy, zero_iff], },
+  rw show y = 1 * y, by rw one_mul,
+  rw show x = (x * y⁻¹) * y, { rw [mul_assoc, inv_mul_cancel hy, mul_one], },
+  iterate 2 {rw [v.map_mul _ y, v'.map_mul _ y]},
+  rw [v.map_one, v'.map_one],
+  split; intro H,
+  { apply mul_le_mul_right',
+    replace hy := v.ne_zero_iff.mpr hy,
+    replace H := le_of_le_mul_right hy H,
+    rwa h at H, },
+  { apply mul_le_mul_right',
+    replace hy := v'.ne_zero_iff.mpr hy,
+    replace H := le_of_le_mul_right hy H,
+    rwa h, },
+end
+
+section supp
+variables  [comm_ring R]
+variables (v : valuation R Γ₀)
+
+/-- The support of a valuation v : R → Γ₀ is the ideal of R where v vanishes. -/
+def supp : ideal R :=
+{ carrier   := {x | v x = 0},
+  zero_mem' := map_zero v,
+  add_mem'  := λ x y hx hy, le_zero_iff.mp $
+    calc v (x + y) ≤ max (v x) (v y) : v.map_add x y
+               ... ≤ 0               : max_le (le_zero_iff.mpr hx) (le_zero_iff.mpr hy),
+  smul_mem' := λ c x hx, calc v (c * x)
+                        = v c * v x : map_mul v c x
+                    ... = v c * 0 : congr_arg _ hx
+                    ... = 0 : mul_zero _ }
+
+@[simp] lemma mem_supp_iff (x : R) : x ∈ supp v ↔ v x = 0 := iff.rfl
+-- @[simp] lemma mem_supp_iff' (x : R) : x ∈ (supp v : set R) ↔ v x = 0 := iff.rfl
+
+/-- The support of a valuation is a prime ideal. -/
+instance : ideal.is_prime (supp v) :=
+⟨λ (h : v.supp = ⊤), one_ne_zero $ show (1 : Γ₀) = 0,
+from calc 1 = v 1 : v.map_one.symm
+        ... = 0   : show (1:R) ∈ supp v, by rw h; trivial,
+ λ x y hxy, begin
+    show v x = 0 ∨ v y = 0,
+    change v (x * y) = 0 at hxy,
+    rw [v.map_mul x y] at hxy,
+    exact eq_zero_or_eq_zero_of_mul_eq_zero hxy
+  end⟩
+
+/-- v(a)=v(a+s) if s ∈ supp(v). -/
+lemma val_add_supp (a s : R) (h : s ∈ supp v) : v (a + s) = v a :=
+begin
+  have aux : ∀ a s, v s = 0 → v (a + s) ≤ v a,
+  { intros a' s' h', refine le_trans (v.map_add a' s') (max_le (le_refl _) _), simp [h'], },
+  apply le_antisymm (aux a s h),
+  calc v a = v (a + s + -s) : by simp
+       ... ≤ v (a + s)      : aux (a + s) (-s) (by rwa ←ideal.neg_mem_iff at h)
+end
+
+/-- If `hJ : J ⊆ supp v` then `on_quot_val hJ` is the induced function on R/J as a function.
+Note: it's just the function; the valuation is `on_quot hJ`. -/
+definition on_quot_val {J : ideal R} (hJ : J ≤ supp v) :
```

Comment:
I don't see `definition` written out in full too often anymore! :D

Reply:
Now it's good old `def` again.